modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -232,9 +246,11 @@ public InternalTransaction begin(HybridTimestampTracker timestampTracker) {
     public InternalTransaction begin(HybridTimestampTracker timestampTracker, boolean readOnly) {
         HybridTimestamp beginTimestamp = clock.now();
         UUID txId = transactionIdGenerator.transactionIdFor(beginTimestamp);
-        updateTxMeta(txId, old -> new TxStateMeta(PENDING, localNodeId, null, null));
+        updateTxMeta(txId, old -> new TxStateMeta(PENDING, localNodeId, null, null, readOnly));
 
         if (!readOnly) {
+            rwTxCounterByCatalogVersion.incrementTxCount(txCatalogVersion(beginTimestamp));
```

Comment:
There is a problem with this approach. As we resolve catalog version by ts without doing a proper schema sync (to avoid waiting), we might get different results when doing increment and decrement, so we might increment for one catalog version, but decrement for a later version, so the transaction might make the index await hang for a long time (till next catalog update arrives).

It seems that we can solve this by resolving the catalog version by the ts (in the same way it's done now) before doing increment and save the catalog version in `TxstateMeta`; then, when doing decrement, use this version from the catalog version.

Reply:
You're right, I'll fix it as you suggested.